src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -127,35 +127,42 @@ private void mapNodeToWorker(List<IWorker<T>> runnables, List<T> freeNodes) {
     if (!RuntimeBehavior.enforceThreadAffinity()) {
       return;
     }
-    int index = 0;
     for (IWorker<T> runnable : runnables) {
-      T freeNode = freeNodes.get(index++);
-      IWorker<T> w = mapping.get(freeNode);
-      if (w != null) {
-        long current = w.getThreadIdToRunOn();
-        runnable.setThreadIdToRunOn(current);
+      for (T freeNode : freeNodes) {
+        IWorker<T> w = mapping.get(freeNode);
+        if (w != null) {
+          long current = w.getThreadIdToRunOn();
+          runnable.setThreadIdToRunOn(current);
+        }
+        mapping.put(freeNode, runnable);
```

Comment:
How do you prevent that the runnable will be mapped on all free nodes?

Reply:
I thought through this a bit more. The only time when that can happen is when there are two or more free nodes (which essentially are dependent methods) that need to use the same thread id as their upstream method. In that case, we will essentially run both the methods in the same thread.

For e.g., if user is saying `parallel=methods` and have the below setup with thread affinity enabled.

* `testa()` - independent
* `testb()` - depends on `testa()`
* `testc()` - depends on `testa()`

This will cause both `testb()` and `testc()` to run on the same thread as `testa()` [ In a way this kind of enforces sequential execution, but that is fine I guess because user wants to have thread affinity, so they already are looking for something like this ]